File: ruby/bubble_sort.rb
def bubble_sort(arr)
  return arr if arr.empty?

  i = 0
  while i < arr.length
    s = 0
    while s < arr.length - 1
      if arr[s] > arr[s + 1]
        arr[s], arr[s + 1] = arr[s + 1], arr[s]
      end
      s += 1
    end
    i += 1
  end
  arr
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: [1, 2, 3, 4]"
  print "=> "
  print bubble_sort([3, 2, 1, 4])

  puts

  puts "Expecting: [1, 2, 3]"
  print "=> "
  print bubble_sort([1, 2, 3])

  puts

  puts "Expecting: []"
  print "=> "
  print bubble_sort([])

  puts

  puts "Expecting: [1, 2, 3]"
  print "=> "
  print bubble_sort([2, 3, 1])

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
# And a written explanation of your solution

File: javascript/bubble_sort.js
function bubbleSort(arr) {
  if (arr.length === 0) {
    return arr;
  } else {
    for (i = 0; i < arr.length; i++) {
      let j = 0;
      while (j < arr.length) {
        if (arr[j] > arr[j + 1]) {
          [arr[j], arr[j + 1]] = [arr[j + 1], arr[j]];
        }
        j++;
      }
    }
  }
  return arr;
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: [1, 2, 3, 4]");
  console.log("=>", bubbleSort([3, 2, 1, 4]));

  console.log("");

  console.log("Expecting: [1, 2, 3]");
  console.log("=>", bubbleSort([1, 2, 3]));

  console.log("");

  console.log("Expecting: []");
  console.log("=>", bubbleSort([]));

  console.log("");

  console.log("Expecting: [1, 2, 3]");
  console.log("=>", bubbleSort([2, 3, 1]));
}

module.exports = bubbleSort;

// Please add your pseudocode to this file
// 1. Use for loop to iterate over Array.
// 2. If value of arr[i + 1] < arr[i], use destructuring: [arr[i], arr[i + 1]] = [arr[i + 1], arr[i]].
// 3. If array is empty, return empty array.
// 4. Once done iterating, return arr.
// And a written explanation of your solution
